pkg/hub/submarineraddonagent/agent.go

```diff
@@ -0,0 +1,189 @@
+func (a *addOnAgent) Manifests(cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn) ([]runtime.Object, error) {
+	objects := []runtime.Object{}
+
+	installNamespace := addon.Spec.InstallNamespace
+	if len(installNamespace) == 0 {
+		installNamespace = defaultInstallationNamespace
+	}
+
+	manifestConfig := struct {
+		KubeConfigSecret      string
+		ClusterName           string
+		AddonInstallNamespace string
+	}{
+		KubeConfigSecret:      fmt.Sprintf("%s-hub-kubeconfig", a.GetAgentAddonOptions().AddonName),
+		AddonInstallNamespace: installNamespace,
+		ClusterName:           cluster.Name,
+	}
+
+	for _, file := range agentDeploymentFiles {
+		raw := assets.MustCreateAssetFromTemplate(file, bindata.MustAsset(filepath.Join("", file)), &manifestConfig).Data
+		object, _, err := genericCodec.Decode(raw, nil, nil)
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, object)
+	}
+	return objects, nil
+}
+
+// GetAgentAddonOptions returns the options of submariner-addon agent
+func (a *addOnAgent) GetAgentAddonOptions() agent.AgentAddonOptions {
+	return agent.AgentAddonOptions{
+		AddonName: addOnName,
+		Registration: &agent.RegistrationOption{
+			CSRConfigurations: agent.KubeClientSignerConfigurations(addOnName, a.agentName),
+			CSRApproveCheck:   a.csrApproveCheck,
+			PermissionConfig:  a.permissionConfig,
+		},
+	}
+}
+
+// To check the addon agent csr, we check
+// 1. if the signer name in csr request is valid.
+// 2. if organization field and commonName field in csr request is valid.
+// 3. if user name in csr is the same as commonName field in csr request.
+func (a *addOnAgent) csrApproveCheck(cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn, csr *certificatesv1.CertificateSigningRequest) bool {
+	if csr.Spec.SignerName != certificatesv1.KubeAPIServerClientSignerName {
+		return false
+	}
+
+	block, _ := pem.Decode(csr.Spec.Request)
+	if block == nil || block.Type != "CERTIFICATE REQUEST" {
+		klog.V(4).Infof("csr %q was not recognized: PEM block type is not CERTIFICATE REQUEST", csr.Name)
+		return false
+	}
+
+	x509cr, err := x509.ParseCertificateRequest(block.Bytes)
+	if err != nil {
+		klog.V(4).Infof("csr %q was not recognized: %v", csr.Name, err)
+		return false
+	}
+
+	requestingOrgs := sets.NewString(x509cr.Subject.Organization...)
+	if requestingOrgs.Len() != 3 {
+		return false
+	}
+
+	if !requestingOrgs.Has("system:authenticated") {
+		return false
+	}
+
+	if !requestingOrgs.Has(fmt.Sprintf("system:open-cluster-management:addon:%s", addOnName)) {
+		return false
+	}
+
+	clusterAddOnGroup := fmt.Sprintf("system:open-cluster-management:cluster:%s:addon:%s", cluster.Name, addOnName)
+	if !requestingOrgs.Has(clusterAddOnGroup) {
+		return false
+	}
+
+	if !strings.HasPrefix(x509cr.Subject.CommonName, clusterAddOnGroup+":agent:") {
+		return false
+	}
+
+	return csr.Spec.Username == x509cr.Subject.CommonName
```

Comment:
in current implementation, i do not think this will be equal. @elgnay is it correct?

Reply:
No. the user name of csr is different with common name in cert. In the current implementation, csr for addon registration is always created with hub-kubeconfig of registration-agent.

Reply:
yes, you are right, I check the csr request common name here